MicroBuild v2: changes to project files and build scripts to enable building in MBv2, and PGO builds of ChakraCore; add nuspec for ChakraCore.

dilijev opened the pull request:
Thanks to @Cellule for helping with the PGO configurations.

Includes initial work on NuGet as requested in issue #85

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/microsoft/chakracore/838)
<!-- Reviewable:end -->

dilijev commented:
@Cellule @tcare @digitalinfinity I'd like to get this in ASAP to unblock internal builds. Can you give a quick review?

tcare commented:
LGTM, please wait for @digitalinfinity's signoff.

Cellule commented:
I saw changes to sign ChakraCore.dll, does ch.exe needs that change as well to be signed ? 

---

Reviewed 41 of 41 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[bin/ChakraCore/ChakraCore.vcxproj, line 80 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFox-z5v9K1RQuYttX0:-KFox-z5v9K1RQuYttX1:1420449256) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/bin/ChakraCore/ChakraCore.vcxproj#L80)):*
Tom and I determined yesterday that this is indeed a mistake. It should be `PGOptimize`

---

*[Build/Microsoft.ChakraCore.nuspec, line 20 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFoxnAkXhfTslLHf6w5:-KFoxnAkXhfTslLHf6w6:-1239145287) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/Microsoft.ChakraCore.nuspec#L20)):*
Why not publish arm binaries as well ?

---

*[Build/scripts/finalize_build.ps1, line 1 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFoyNEKIbMhtWzLkuPt:-KFoyNEKIbMhtWzLkuPu:-1171616130) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/finalize_build.ps1#L1)):*
When is this run exactly ? Is it once for all flavors or do all flavors run this ?

---

*[Build/scripts/post_pgi.cmd, line 6 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFozT-fP6X8oc36qmbd:-KFozT-fP6X8oc36qmbe:1905075566) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/post_pgi.cmd#L6)):*
Update or remove this comment

---

*[Build/scripts/post_pgo.cmd, line 6 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFozW9_jCFX_dz4kJw6:-KFozW9_jCFX_dz4kJw7:-1685473906) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/post_pgo.cmd#L6)):*
update or remove this comment

---

*[Build/scripts/pre_pgi.cmd, line 6 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFozvJYhKuh-NLYbREl:-KFozvJZzkvjsrB58hhC:-1516705522) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/pre_pgi.cmd#L6)):*
update or remove comment

---

*[Build/scripts/pre_pgo.cmd, line 6 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFozy0wTvvNt6QRSieE:-KFozy0wTvvNt6QRSieF:627510574) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/pre_pgo.cmd#L6)):*
Update or remove comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/838#-:-KFoxHA2JTi7dpEh5Zpi:-1614899190)*
<!-- Sent from Reviewable.io -->

digitalinfinity commented on Build/scripts/pogo_training.ps1:
```diff
@@ -0,0 +1,45 @@
+param (
```
Nit: Can we move all the pogo related scripts to its own folder?

dilijev replied:
I don't see why not. I'll move them.

digitalinfinity commented on Build/scripts/pogo_training.ps1:
```diff
@@ -0,0 +1,45 @@
+    } elseif ($arch -eq "arm") {
+        $vcarchpath = "arm"
+    }
+    $pgoSrcDll = Join-Path ${env:ProgramFiles(x86)} (Join-Path (Join-Path "Microsoft Visual Studio 14.0\VC\bin" $vcarchpath) "pgort140.dll")
```
Consider making the VC bin path and pgo dll parameters?

Cellule replied:
For this one I have a better fix using property sheet. It will be done at a later point

dilijev replied:
Good idea.

dilijev commented:
This nuspec will be used for MyGet packages and should include a ch.exe because we are interested in people being able to easily try the latest features. We will have a separate spec for NuGet which only includes the signed ChakraCore.dll files because it is intended for people to take a dependency on and including the ch.exe host is not productive there.

As for signing ch.exe for the MyGet package, I will follow up with @edmaurer about our exact plan here. It will be easy to add signing to additional binaries if/when we determine we would like to.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[bin/ChakraCore/ChakraCore.vcxproj, line 80 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFox-z5v9K1RQuYttX0:-KFp2Pe1EdOS6wbdr2MD:397799397) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/bin/ChakraCore/ChakraCore.vcxproj#L80)):*
Thanks. I'm fixing this.
The other location has `<LinkTimeCodeGeneration Condition="'$(POGO_TYPE)'=='PGO'">PGOptimization</LinkTimeCodeGeneration>`
Is it `PGOptimization` or `PGOptimize`?

---

*[Build/Microsoft.ChakraCore.nuspec, line 20 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFoxnAkXhfTslLHf6w5:-KFp2mH06QnfqvOGk0tF:425813457) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/Microsoft.ChakraCore.nuspec#L20)):*
We don't officially support an ARM build at this time for various reasons including because we don't have full test coverage. We should talk about this offline.

---

*[Build/scripts/finalize_build.ps1, line 1 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFoyNEKIbMhtWzLkuPt:-KFp496P8cX0SOjK6bNA:-141941339) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/finalize_build.ps1#L1)):*
Every flavor runs this. It only deals with creating and deleting files that are used to track the status of just this flavor. There will be a different release configuration that handles running any tasks that involve composing the whole build.

---

*[Build/scripts/post_pgi.cmd, line 6 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFozT-fP6X8oc36qmbd:-KFp6CY1DpzrAGtBOhGQ:553742513) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/post_pgi.cmd#L6)):*
This is more for documentation for me, so I've removed it from all files (and wrote down the corrected version.)

---

*[Build/scripts/post_pgo.cmd, line 6 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFozW9_jCFX_dz4kJw6:-KFp6JOgIIfiVT3zQ-xY:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/post_pgo.cmd#L6)):*
Done.

---

*[Build/scripts/pre_pgi.cmd, line 6 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFozvJYhKuh-NLYbREl:-KFp6K8EP5IJQA9CEEU4:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/pre_pgi.cmd#L6)):*
Done.

---

*[Build/scripts/pre_pgo.cmd, line 6 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFozy0wTvvNt6QRSieE:-KFp6KTx2JcTVWlhs8vO:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/pre_pgo.cmd#L6)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/838#-:-KFp0r-IgX_nNKqKIXMK:388009972)*
<!-- Sent from Reviewable.io -->

digitalinfinity commented on Build/scripts/init_build.ps1:
```diff
@@ -0,0 +1,153 @@
+# Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+#-------------------------------------------------------------------------------------------------------
+
+param (
```
Consider adding a comment here on what this script's purpose is and where it'll be called

digitalinfinity commented:
:+1:

Cellule commented:
I meant the change you did in ChakraCore.vcxproj
```
</ItemGroup>
  <!-- Conditionally include the group of files to sign because MBv2 knows what to do with this but the old build definition gets confused. -->
  <ItemGroup Condition="'$(VSO_MICROBUILD_V2)'=='True'">
    <FilesToSign Include="$(OutDir)\ChakraCore.dll">
      <Authenticode>Microsoft</Authenticode>
    </FilesToSign>
  </ItemGroup>
```
If this is only for MyGet packages then nevermind, but it looks like an instruction to tell microbuildv2 to sign this binary. 

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[bin/ChakraCore/ChakraCore.vcxproj, line 80 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFox-z5v9K1RQuYttX0:-KFp991zk10pI4QeU3c_:-111086811) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/bin/ChakraCore/ChakraCore.vcxproj#L80)):*
I wasn't able to confirm, but I think it fine as it is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/838#-:-KFp8r8tj24Ln9ky3qAW:2095277235)*
<!-- Sent from Reviewable.io -->

dilijev commented:
Yes, that's right. That ItemGroup tells MBv2 to sign the files included in that group. After discussing offline with @digitalinfinity we've decided that it's best to sign all binaries we release publically in any form as per corp guidelines. The set of files we include in which package is a different issue, and we will probably stick to including both files in MyGet and only ChakraCore.dll/.pdb in the NuGet.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[bin/ChakraCore/ChakraCore.vcxproj, line 80 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFox-z5v9K1RQuYttX0:-KFpH6Pm9zinb8UTiOaX:-767902398) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/bin/ChakraCore/ChakraCore.vcxproj#L80)):*
Okay I'm changing `PGOnstrument` to `PGOptimization` here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/838#-:-KFpGa1zexIPYzInAQ3i:-895543181)*
<!-- Sent from Reviewable.io -->

Cellule commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[bin/ChakraCore/ChakraCore.vcxproj, line 80 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFox-z5v9K1RQuYttX0:-KFpHtlX9cm2UEG0-es1:254063984) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/bin/ChakraCore/ChakraCore.vcxproj#L80)):*
*`PGOptimize`* not `PGOptimization`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/838)*
<!-- Sent from Reviewable.io -->

dilijev commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[bin/ChakraCore/ChakraCore.vcxproj, line 80 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFox-z5v9K1RQuYttX0:-KFpWVa1zYEXR1Rw5OnG:-1328441352) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/bin/ChakraCore/ChakraCore.vcxproj#L80)):*
Okay I'm changing this one to `PGOptimize`. Should the other one stay as `PGOptimization`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/838)*
<!-- Sent from Reviewable.io -->

dilijev commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[Build/scripts/init_build.ps1, line 6 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFowYvF5acNxjQ8wrsJ-r1-6:-KFpWbwyMQXcXqIeq4Kb:-499146009) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/init_build.ps1#L6)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/838)*
<!-- Sent from Reviewable.io -->

Cellule commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[bin/ChakraCore/ChakraCore.vcxproj, line 80 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFox-z5v9K1RQuYttX0:-KFpdB-cGBeEXYb9N4KJ:-1001801788) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/bin/ChakraCore/ChakraCore.vcxproj#L80)):*
I need to verify, but leave it like this at this time

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/838)*
<!-- Sent from Reviewable.io -->

dilijev commented:
@digitalinfinity @Cellule 

Added a new commit to address comments:
Addressed review comments: pgo script folder, sign ch.exe, generalize parameters, create packages.config files dynamically, doc comments.

Does everything look alright?

Cellule commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 17 of 38 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/838#-:-KFpjbYnaipeorfwcf54:1417088721)*
<!-- Sent from Reviewable.io -->

dilijev commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[Build/scripts/pogo_training.ps1, line 24 \[r1\]](https://reviewable.io:443/reviews/microsoft/chakracore/838#-KFowYvF5acNxjQ8wrsK-r1-24:-KFpm_K0roYtu8lqvLcS:-680727111) ([raw file](https://github.com/microsoft/chakracore/blob/2f13802dcd09d3ea58f7e381e3c8a4de088ca9ca/Build/scripts/pogo_training.ps1#L24)):*
Okay, Michael. I'll keep an eye out for that change. For now, this will do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/microsoft/chakracore/838)*
<!-- Sent from Reviewable.io -->
